File: src/event/src/lib/client-error-reporter.ts
import { tracingFetch } from './tracing-fetch';

type ClientErrorSeverity = 'error' | 'fatal';

type ClientErrorPayload = {
  platform: 'web';
  sourceService?: 'framefast-event';
  errorType: string;
  message: string;
  stack?: string;
  handled?: boolean;
  severity?: ClientErrorSeverity;
  url?: string;
  route?: string;
  release?: string;
  userAgent?: string;
  metadata?: Record<string, string | number | boolean>;
};

const MAX_MESSAGE = 2000;
const MAX_STACK = 16000;
const MAX_URL = 2048;
const MAX_ROUTE = 512;
let handlersInstalled = false;

function truncate(value: string, max: number): string {
  return value.length > max ? value.slice(0, max) : value;
}

function parseErrorLike(reason: unknown): { message: string; stack?: string } {
  if (reason instanceof Error) {
    return {
      message: reason.message || reason.name || 'Unknown error',
      stack: reason.stack,
    };
  }
  if (typeof reason === 'string') {
    return { message: reason };
  }
  try {
    return { message: JSON.stringify(reason) };
  } catch {
    return { message: String(reason) };
  }
}

export async function reportClientError(payload: ClientErrorPayload): Promise<void> {
  const safePayload = {
    ...payload,
    sourceService: 'framefast-event' as const,
    message: truncate(payload.message || 'Unknown error', MAX_MESSAGE),
    stack: payload.stack ? truncate(payload.stack, MAX_STACK) : undefined,
    url: payload.url ? truncate(payload.url, MAX_URL) : undefined,
    route: payload.route ? truncate(payload.route, MAX_ROUTE) : undefined,
    severity: payload.severity ?? 'error',
    handled: payload.handled ?? false,
    release: import.meta.env.MODE,
    userAgent: navigator.userAgent,
  };

  try {
    await tracingFetch(`${import.meta.env.VITE_API_URL}/observability/client-errors`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      keepalive: true,
      body: JSON.stringify(safePayload),
    });
  } catch {
    // Never throw from error reporter path.
  }
}

export function installGlobalErrorHandlers(): void {
  if (handlersInstalled) return;
  handlersInstalled = true;

  window.addEventListener('error', (event) => {
    const details = parseErrorLike(event.error ?? event.message);
    void reportClientError({
      platform: 'web',
      sourceService: 'framefast-event',
      errorType: 'window_error',
      message: details.message,
      stack: details.stack,
      handled: false,
      severity: 'error',
      url: window.location.href,
      route: window.location.pathname,
      metadata: {
        lineno: event.lineno || 0,
        colno: event.colno || 0,
      },
    });
  });

  window.addEventListener('unhandledrejection', (event) => {
    const details = parseErrorLike(event.reason);
    void reportClientError({
      platform: 'web',
      sourceService: 'framefast-event',
      errorType: 'unhandled_rejection',
      message: details.message,
      stack: details.stack,
      handled: false,
      severity: 'error',
      url: window.location.href,
      route: window.location.pathname,
    });
  });
}

File: studio/FrameFast/API/ClientErrorReporter.swift
//
//  ClientErrorReporter.swift
//  FrameFast
//

import Foundation

enum ClientErrorReporter {
    private struct Payload: Encodable {
        let platform: String
        let sourceService: String
        let errorType: String
        let message: String
        let stack: String?
        let handled: Bool
        let severity: String
        let route: String?
        let release: String?
        let traceparent: String?
        let baggage: String?
        let metadata: [String: String]?
    }

    static func reportNonFatal(
        baseURL: String,
        request: URLRequest?,
        errorType: String,
        message: String,
        error: Error?,
        metadata: [String: String] = [:]
    ) {
        Task.detached(priority: .utility) {
            await send(
                baseURL: baseURL,
                request: request,
                errorType: errorType,
                message: message,
                error: error,
                metadata: metadata
            )
        }
    }

    private static func send(
        baseURL: String,
        request: URLRequest?,
        errorType: String,
        message: String,
        error: Error?,
        metadata: [String: String]
    ) async {
        guard let url = URL(string: "\(baseURL)/observability/client-errors") else { return }

        let normalizedMessage = String(message.prefix(2_000))
        let stackText = error.map { String(String(reflecting: $0).prefix(16_000)) }
        let route = request?.url?.path
        let release = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        let traceparent = request?.value(forHTTPHeaderField: "traceparent")
        let baggage = request?.value(forHTTPHeaderField: "baggage")
        let payload = Payload(
            platform: "ios",
            sourceService: "framefast-ios",
            errorType: String(errorType.prefix(64)),
            message: normalizedMessage,
            stack: stackText,
            handled: true,
            severity: "error",
            route: route,
            release: release,
            traceparent: traceparent,
            baggage: baggage,
            metadata: metadata.isEmpty ? nil : metadata
        )

        var post = URLRequest(url: url)
        post.httpMethod = "POST"
        post.setValue("application/json", forHTTPHeaderField: "Content-Type")
        post.timeoutInterval = 5

        do {
            post.httpBody = try JSONEncoder().encode(payload)
            _ = try await URLSession.shared.data(for: post)
        } catch {
            // Never throw from reporter path.
        }
    }
}
